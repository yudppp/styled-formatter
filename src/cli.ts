#!/usr/bin/env node

import { program } from "commander";
import fs from "fs";
import { globSync } from "glob";
import kleur from "kleur";
import { diffLines } from "diff";
import { formatTemplateLiterals } from "./index";

/**
 * Define the type of indentation
 */
type IndentType = "tab" | number;

/**
 * Default indentation setting
 */
const DEFAULT_INDENT = 2;

/**
 * CLI entry point
 */
program
	.version("1.0.0")
	.description("A CLI for formatting styled-components template literals")
	.option("--fix", "Automatically fix files")
	.option(
		"--indent <type>",
		"Indentation type (tab or number of spaces, default: 2)",
		parseIndentOption,
	)
	.arguments("[files...]")
	.action(
		async (files: string[], options: { fix: boolean; indent: IndentType }) => {
			const filePatterns = files.length > 0 ? files : ["**/*.ts", "**/*.tsx"];
			const shouldFix = options.fix || false;
			const indentType = options.indent || DEFAULT_INDENT;

			// Resolve file paths
			const resolvedFiles = resolveFiles(filePatterns);

			if (resolvedFiles.length === 0) {
				console.log(kleur.yellow("No files found."));
				process.exit(0);
			}

			let hasErrors = false;

			// Process each file asynchronously
			for (const filePath of resolvedFiles) {
				try {
					const originalCode = fs.readFileSync(filePath, "utf-8");
					const formattedCode = await formatTemplateLiterals(originalCode, {
						indent: indentType,
					});
					if (originalCode !== formattedCode) {
						if (shouldFix) {
							// Write the fixed code back to the file
							fs.writeFileSync(filePath, formattedCode, "utf-8");
							console.log(kleur.green(`Fixed: ${filePath}`));
						} else {
							// Display differences
							hasErrors = true;
							console.log(kleur.bold(`\nDiff: ${filePath}`));
							const diff = diffLines(originalCode, formattedCode);
							diff.forEach((part) => {
								const color = part.added
									? kleur.green
									: part.removed
										? kleur.red
										: kleur.gray;
								process.stdout.write(color(part.value));
							});
							console.log("\n");
						}
					} else if (!shouldFix) {
						console.log(kleur.green(`No changes needed: ${filePath}`));
					}
				} catch (error) {
					hasErrors = true;
					console.error(kleur.red(`Error processing file: ${filePath}`));
					console.error(error);
				}
			}

			// Final message and exit status
			if (!shouldFix && hasErrors) {
				console.log(kleur.red("\nFormatting issues found."));
				process.exit(1);
			} else {
				console.log(kleur.green("\nAll files are formatted correctly."));
				process.exit(0);
			}
		},
	);

program.parse(process.argv);

/**
 * Function to parse the indentation option
 * @param value - Value received from the command line
 * @returns Parsed indentation setting
 */
function parseIndentOption(value: string): IndentType {
	if (value === "tab") {
		return "tab";
	}

	const numSpaces = parseInt(value, 10);
	if (isNaN(numSpaces) || numSpaces < 0) {
		throw new Error("Indent must be 'tab' or a positive number");
	}

	return numSpaces;
}

/**
 * Resolves file patterns and returns an array of file paths
 * @param patterns - Glob patterns to resolve
 * @returns Array of resolved file paths
 */
function resolveFiles(patterns: string[]): string[] {
	const resolvedFiles: string[] = [];
	patterns.forEach((pattern) => {
		const files = globSync(pattern, {
			absolute: true,
			ignore: ["node_modules/**", "dist/**"],
		});
		resolvedFiles.push(...files);
	});
	return resolvedFiles;
}
