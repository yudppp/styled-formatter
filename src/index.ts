import { parse } from "@babel/parser";
import traverse, { NodePath } from "@babel/traverse";
import * as t from "@babel/types";
import prettier from "prettier";

/**
 * Format the specified content using Prettier
 * @param code Code to be formatted
 * @param indent Indentation type (tab or number of spaces)
 * @returns Formatted code
 */
export const formatWithContext = async (
	code: string,
	indent: "tab" | number,
): Promise<string> => {
	try {
		return prettier.format(code, {
			semi: false,
			parser: "css",
			useTabs: indent === "tab",
			tabWidth: indent === "tab" ? 1 : indent,
		});
	} catch (error) {
		console.error("Failed to format code with Prettier:", error);
		return code;
	}
};

export type FormatOption = {
	indent: "tab" | number;
};

/**
 * Determines if a given node is a styled-components tag
 * @param tag The AST node to check
 * @returns boolean indicating if the node is a styled-components tag
 */
const isStyledComponentTag = (tag: t.Expression): boolean => {
	if (t.isMemberExpression(tag)) {
		// styled.div
		return t.isIdentifier(tag.object, { name: "styled" });
	}

	if (t.isCallExpression(tag)) {
		const callee = tag.callee;
		return (
			t.isIdentifier(callee, { name: "styled" }) ||
			(t.isMemberExpression(callee) &&
				t.isIdentifier(callee.object, { name: "styled" }))
		);
	}

	return false;
};

/**
 * Creates a unique placeholder for expressions
 * @param counter Current counter value
 * @param before Character before the expression
 * @param after Character after the expression
 * @returns Generated placeholder
 */
const createPlaceholder = (
	counter: number,
	beforeText: string,
	afterText: string,
): string => {
	const base = `__EXPR_${counter}__`;
	const isConnectChar = /^[a-zA-Z:;]/.test(afterText.slice(0, 1));
	if (isConnectChar) {
		return `\$${base}`;
	}
	return beforeText.trimEnd().slice(-1) === ":" ? `\$${base}` : `/* ${base} */`;
};

/**
 * Filters an array to include only Expression nodes
 * @param items Array containing Expression or TSType nodes
 * @returns Array containing only Expression nodes
 */
const filterExpressions = (
	items: (t.Expression | t.TSType)[],
): t.Expression[] => {
	return items.filter((item): item is t.Expression => t.isExpression(item));
};

/**
 * Replaces expressions and comments in template literals with placeholders
 * @param quasis Template elements
 * @param expressions Embedded expressions (filtered to Expression[])
 * @param originalCode Original code string
 * @returns Object containing the modified string and mappings for restoration
 */
const replacePlaceholders = (
	quasis: t.TemplateElement[],
	expressions: t.Expression[],
	originalCode: string,
) => {
	let placeholderCounter = 0;
	let commentCounter = 0;
	const expressionMap: Record<string, string> = {};
	const commentMap: Record<string, string> = {};

	const formattedParts = quasis.map((quasi, index) => {
		let text = quasi.value.cooked || quasi.value.raw;

		if (index < expressions.length) {
			const nextQuasi = quasis[index + 1];
			const beforeText = text;
			const afterText = nextQuasi?.value?.cooked ?? nextQuasi?.value?.raw ?? "";
			const placeholder = createPlaceholder(
				placeholderCounter++,
				beforeText,
				afterText,
			);
			const expr = expressions[index];

			if (expr.start === undefined || expr.end === undefined) {
				throw new Error("Expression does not have start or end position");
			}

			expressionMap[placeholder] = originalCode.slice(
				expr.start ?? 0,
				expr.end ?? 0,
			);
			text += placeholder;
		}

		// Replace inline comments
		const commentRegex1 = /\/\*([\s\S]*?)\*\//gm;
		text = text.replace(commentRegex1, (match) => {
			const commentPlaceholder = `/* __COMMENT_${commentCounter++}__ */`;
			commentMap[commentPlaceholder] = match;
			return commentPlaceholder;
		});
		const commentRegex2 = /\/\/.*$/gm;
		text = text.replace(commentRegex2, (match) => {
			const commentPlaceholder = `/* __COMMENT_${commentCounter++}__ */`;
			commentMap[commentPlaceholder] = match;
			return commentPlaceholder;
		});
		return text;
	});

	return {
		concatenated: formattedParts.join(""),
		expressionMap,
		commentMap,
	};
};

/**
 * Restores expressions and comments in the formatted template literal
 * @param formatted Formatted template string with placeholders
 * @param expressionMap Mapping of expression placeholders to original expressions
 * @param commentMap Mapping of comment placeholders to original comments
 * @param indentChar Indentation characters
 * @returns Restored template literal string
 */
const restorePlaceholders = (
	formatted: string,
	expressionMap: Record<string, string>,
	commentMap: Record<string, string>,
	indentChar: string,
): string => {
	if (formatted.trim() === "") return "``";
	let restored =
		"`\n" +
		formatted
			.split("\n")
			.map((line) => (line === "" ? "" : `${indentChar}${line}`))
			.join("\n") +
		"`";
	for (const [placeholder, comment] of Object.entries(commentMap)) {
		restored = restored.replace(placeholder, comment);
	}

	for (const [placeholder, expr] of Object.entries(expressionMap)) {
		restored = restored.replace(placeholder, `\${${expr}}`);
	}

	return restored;
};

/**
 * Formats template literals within styled-components.
 * Replaces embedded expressions (${...}) with placeholders before formatting
 * and restores them afterward. The replacement method is determined based on
 * surrounding characters.
 * @param code TypeScript code to be formatted
 * @param option Formatting options
 * @returns Formatted code
 */
export const formatTemplateLiterals = async (
	code: string,
	option: FormatOption = { indent: 2 },
): Promise<string> => {
	const { indent } = option;
	const indentChar = indent === "tab" ? "\t" : " ".repeat(indent);

	try {
		const ast = parse(code, {
			sourceType: "module",
			plugins: ["typescript", "jsx"],
			ranges: true,
			tokens: true,
		});

		const taggedTemplateExpressions: t.TaggedTemplateExpression[] = [];

		traverse(ast, {
			TaggedTemplateExpression(path: NodePath<t.TaggedTemplateExpression>) {
				if (
					isStyledComponentTag(path.node.tag) &&
					path.node.quasi.quasis.length > 0
				) {
					taggedTemplateExpressions.push(path.node);
				}
			},
		});

		const replacements: Array<{ start: number; end: number; content: string }> =
			[];

		for (const node of taggedTemplateExpressions) {
			const { quasis, expressions } = node.quasi;

			const pureExpressions = filterExpressions(expressions);
			if (node.quasi.start !== null && node.quasi.end !== null) {
				const { concatenated, expressionMap, commentMap } = replacePlaceholders(
					quasis,
					pureExpressions,
					code,
				);

				const formatted = await formatWithContext(concatenated, indent);

				const restored = restorePlaceholders(
					formatted,
					expressionMap,
					commentMap,
					indentChar,
				);

				replacements.push({
					start: node.quasi.start ?? 0,
					end: node.quasi.end ?? 0,
					content: restored,
				});
			} else {
				console.warn(
					`Skipped formatting for a node with null start or end: ${node.type}`,
				);
			}
		}

		// Apply replacements from back to front to prevent position shifts
		const sortedReplacements = replacements.sort((a, b) => b.start - a.start);
		let updatedCode = code;

		for (const { start, end, content } of sortedReplacements) {
			updatedCode =
				updatedCode.slice(0, start) + content + updatedCode.slice(end);
		}

		return updatedCode;
	} catch (error) {
		console.error("Failed to format template literals:", error);
		console.error("Input data:", code);
		return code;
	}
};
