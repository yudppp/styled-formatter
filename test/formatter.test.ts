import { formatTemplateLiterals, FormatOption } from "../src/";

describe("formatTemplateLiterals", () => {

    describe("Indentation options", () => {
        it("should format with custom space indentation", async () => {
            const input = `
import styled from "styled-components";

const Button = styled.button\`
color: red;
   &:hover {
     background: blue;
   }
\`;
`;

            const expectedOutput = `
import styled from "styled-components";

const Button = styled.button\`
    color: red;
    &:hover {
        background: blue;
    }
\`;
`;

            const option: FormatOption = { indent: 4 };
            const result = await formatTemplateLiterals(input, option);
            expect(result).toBe(expectedOutput);
        });

        it("should format with tab indentation", async () => {
            const input = `
import styled from "styled-components";

const Button = styled.button\`
  color: red;
   &:hover {
        background: blue;
    }
\`;
`;

            const expectedOutput = `
import styled from "styled-components";

const Button = styled.button\`
\tcolor: red;
\t&:hover {
\t\tbackground: blue;
\t}
\`;
`;

            const option: FormatOption = { indent: "tab" };
            const result = await formatTemplateLiterals(input, option);
            expect(result).toBe(expectedOutput);
        });

        it("should default to 2 spaces indentation when no option is provided", async () => {
            const input = `
import styled from "styled-components";

const Button = styled.button\`
color: red;
   &:hover {
     background: blue;
   }
\`;
`;

            const expectedOutput = `
import styled from "styled-components";

const Button = styled.button\`
  color: red;
  &:hover {
    background: blue;
  }
\`;
`;

            const result = await formatTemplateLiterals(input);
            expect(result).toBe(expectedOutput);
        });

        it("should handle mixed indentation options", async () => {
            const input = `
import styled from "styled-components";

const Container = styled.div\`
color: green;
    padding: 10px;
\tmargin: 5px;
\`;
`;

            const expectedOutput = `
import styled from "styled-components";

const Container = styled.div\`
    color: green;
    padding: 10px;
    margin: 5px;
\`;
`;

            const option: FormatOption = { indent: 4 };
            const result = await formatTemplateLiterals(input, option);
            expect(result).toBe(expectedOutput);
        });
    });

    describe("File format support", () => {
        it("should format `.jsx` files correctly", async () => {
            const input = `
import styled from "styled-components";

const Button = styled.button\`
color: red;
   &:hover {
     background: blue;
   }
\`;
`;

            const expectedOutput = `
import styled from "styled-components";

const Button = styled.button\`
  color: red;
  &:hover {
    background: blue;
  }
\`;
`;

            const option: FormatOption = { indent: 2 };
            const result = await formatTemplateLiterals(input, option);
            expect(result).toBe(expectedOutput);
        });
    });

    describe("Error handling", () => {
        it("should handle malformed template literals gracefully", async () => {
            const input = `
import styled from "styled-components";

const Button = styled.button\`
   color: red;
   &:hover {
        background: blue;
    // Missing closing brace
\`;
`;

            const result = await formatTemplateLiterals(input);
            // Should return original input when formatting fails
            expect(result).toBe(input);
        });

        it("should handle invalid syntax in expressions", async () => {
            const input = `
import styled from "styled-components";

const Button = styled.button\`
    color: \${=> invalid syntax};
\`;
`;

            const result = await formatTemplateLiterals(input);
            // Should return original input when parsing fails
            expect(result).toBe(input);
        });

        it("should handle empty template literals", async () => {
            const input = `
import styled from "styled-components";

const Empty = styled.div\`\`;
`;

            const expectedOutput = `
import styled from "styled-components";

const Empty = styled.div\`\`;
`;

            const result = await formatTemplateLiterals(input);
            expect(result).toBe(expectedOutput);
        });
    });

    describe("Complex scenarios", () => {
        it("should handle mixed inline and block comments", async () => {
            const input = `
import styled from "styled-components";

const Button = styled.button\`
    color: red; // Inline comment
    /* Block comment */
    background: blue; // Another inline comment
    /* Multi-line
       block comment */
    padding: 10px;
\`;
`;

            const expectedOutput = `
import styled from "styled-components";

const Button = styled.button\`
  color: red; // Inline comment
  /* Block comment */
  background: blue; // Another inline comment
  /* Multi-line
       block comment */
  padding: 10px;
\`;
`;

            const result = await formatTemplateLiterals(input);
            expect(result).toBe(expectedOutput);
        });

        it("should handle multiple expressions in a single line", async () => {
            const input = `
import styled from "styled-components";

const Button = styled.button\`
    color: \${props => props.primary ? "red" : "blue"}; background: \${props => props.bg}; padding: \${props => props.size}px;
\`;
`;

            const expectedOutput = `
import styled from "styled-components";

const Button = styled.button\`
  color: \${props => props.primary ? "red" : "blue"};
  background: \${props => props.bg};
  padding: \${props => props.size}px;
\`;
`;

            const result = await formatTemplateLiterals(input);
            expect(result).toBe(expectedOutput);
        });

        it("should handle pseudo-selectors and media queries", async () => {
            const input = `
import styled from "styled-components";

const Button = styled.button\`
    color: red;
    &:hover, &:focus {
        background: blue;
    }
    @media (max-width: \${props => props.theme.breakpoints.mobile}) {
        padding: 8px;
        &:hover {
            background: green;
        }
    }
\`;
`;

            const expectedOutput = `
import styled from "styled-components";

const Button = styled.button\`
  color: red;
  &:hover,
  &:focus {
    background: blue;
  }
  @media (max-width: \${props => props.theme.breakpoints.mobile}) {
    padding: 8px;
    &:hover {
      background: green;
    }
  }
\`;
`;

            const result = await formatTemplateLiterals(input);
            expect(result).toBe(expectedOutput);
        });

        it("should handle nested selectors and complex nesting", async () => {
            const input = `
import styled from "styled-components";

const Nav = styled.nav\`
    display: flex;
    ul {
        list-style: none;
        li {
            margin: 0 10px;
            a {
                text-decoration: none;
                &:hover {
                    color: green;
                }
            }
        }
    }
\`;
`;

            const expectedOutput = `
import styled from "styled-components";

const Nav = styled.nav\`
  display: flex;
  ul {
    list-style: none;
    li {
      margin: 0 10px;
      a {
        text-decoration: none;
        &:hover {
          color: green;
        }
      }
    }
  }
\`;
`;

            const result = await formatTemplateLiterals(input);
            expect(result).toBe(expectedOutput);
        });

        it("should handle interpolation with nested template literals", async () => {
            const input = `
import styled from "styled-components";

const dynamicStyles = props => \`
    color: \${props.color};
    background: \${props.background};
\`;

const Button = styled.button\`
    \${dynamicStyles}
    padding: 10px;
\`;
`;

            const expectedOutput = `
import styled from "styled-components";

const dynamicStyles = props => \`
    color: \${props.color};
    background: \${props.background};
\`;

const Button = styled.button\`
  \${dynamicStyles}
  padding: 10px;
\`;
`;

            const result = await formatTemplateLiterals(input);
            expect(result).toBe(expectedOutput);
        });
    });

    describe("Edge cases", () => {
        it("should handle template literals with only expressions", async () => {
            const input = `
import styled from "styled-components";

const Button = styled.button\`\${props => props.customStyles}\`;
`;

            const expectedOutput = `
import styled from "styled-components";

const Button = styled.button\`
  \${props => props.customStyles}
\`;
`;

            const result = await formatTemplateLiterals(input);
            expect(result).toBe(expectedOutput);
        });

        it("should handle multiple sequential expressions", async () => {
            const input = `
import styled from "styled-components";

const Button = styled.button\`
    \${props => props.margin && 'margin: 8px;'}
    \${props => props.padding && 'padding: 16px;'}
    \${props => props.customStyles}
\`;
`;

            const expectedOutput = `
import styled from "styled-components";

const Button = styled.button\`
  \${props => props.margin && 'margin: 8px;'}
  \${props => props.padding && 'padding: 16px;'}
  \${props => props.customStyles}
\`;
`;

            const result = await formatTemplateLiterals(input);
            expect(result).toBe(expectedOutput);
        });

        it("should handle empty lines and excessive whitespace", async () => {
            const input = `
import styled from "styled-components";


const Button = styled.button\`


    color: red;



    &:hover {
            background: blue;
    }


\`;
`;

            const expectedOutput = `
import styled from "styled-components";


const Button = styled.button\`
  color: red;

  &:hover {
    background: blue;
  }
\`;
`;

            const result = await formatTemplateLiterals(input);
            expect(result).toBe(expectedOutput);
        });

        it("should handle large template literals efficiently", async () => {
            const input = `
import styled from "styled-components";

const LargeComponent = styled.div\`
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: space-between;
    padding: 20px;
    margin: 10px;
    background-color: #f0f0f0;
    border: 1px solid #ccc;
    border-radius: 5px;
    box-shadow: 0 2px 4px rgba(0,0,0,0.1);
    font-size: 16px;
    color: #333;
    &:hover {
        background-color: #e0e0e0;
    }
    @media (max-width: 768px) {
        flex-direction: row;
        padding: 10px;
    }
\`;
`;

            const expectedOutput = `
import styled from "styled-components";

const LargeComponent = styled.div\`
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: space-between;
  padding: 20px;
  margin: 10px;
  background-color: #f0f0f0;
  border: 1px solid #ccc;
  border-radius: 5px;
  box-shadow: 0 2px 4px rgba(0, 0, 0, 0.1);
  font-size: 16px;
  color: #333;
  &:hover {
    background-color: #e0e0e0;
  }
  @media (max-width: 768px) {
    flex-direction: row;
    padding: 10px;
  }
\`;
`;

            const result = await formatTemplateLiterals(input);
            expect(result).toBe(expectedOutput);
        });

        it("should handle template literals with single line styles", async () => {
            const input = `
import styled from "styled-components";

const Inline = styled.div\`color: red;\`;
`;

            const expectedOutput = `
import styled from "styled-components";

const Inline = styled.div\`
  color: red;
\`;
`;

            const result = await formatTemplateLiterals(input);
            expect(result).toBe(expectedOutput);
        });
    });

    describe("Integration scenarios", () => {
        it("should work correctly with existing Prettier configurations", async () => {
            const input = `
import styled from "styled-components";

const Button = styled.button\`color:red;\`;
`;

            const expectedOutput = `
import styled from "styled-components";

const Button = styled.button\`
  color: red;
\`;
`;

            const option: FormatOption = { indent: 2 };
            const result = await formatTemplateLiterals(input, option);
            expect(result).toBe(expectedOutput);
        });

        it("should not alter non-styled-components template literals", async () => {
            const input = `
const message = \`Hello, World!\`;
`;

            const expectedOutput = `
const message = \`Hello, World!\`;
`;

            const result = await formatTemplateLiterals(input);
            expect(result).toBe(expectedOutput);
        });

        it("should handle multiple styled-components in a single file", async () => {
            const input = `
import styled from "styled-components";

const Button = styled.button\`
color: red;
\`;

const Container = styled.div\`
display: flex;
\`;
`;

            const expectedOutput = `
import styled from "styled-components";

const Button = styled.button\`
  color: red;
\`;

const Container = styled.div\`
  display: flex;
\`;
`;

            const result = await formatTemplateLiterals(input);
            expect(result).toBe(expectedOutput);
        });
    });

    describe("Performance tests", () => {
        it("should format large codebases within acceptable time", async () => {
            const input = `
import styled from "styled-components";

const Component1 = styled.div\`
    color: red;
    background: blue;
\`;

const Component2 = styled.button\`
    padding: 10px;
    margin: 5px;
\`;

// Repeat similar components to simulate a large codebase
${Array.from({ length: 100 }, (_, i) => `
const Component${i + 3} = styled.span\`
    font-size: ${12 + i}px;
    line-height: ${1.5 + i * 0.1};
\`;
`).join('\n')}
`;

            const startTime = Date.now();
            const result = await formatTemplateLiterals(input);
            const endTime = Date.now();
            const duration = endTime - startTime;

            // Assuming the formatter should process within 2 seconds for this input
            expect(duration).toBeLessThan(2000);

            // Basic check to ensure formatting was applied
            expect(result).toContain("font-size: ");
            expect(result).toContain("line-height: ");
        }, 10000); // Setting Jest timeout to 10 seconds for this test
    });

    describe("Snapshot tests", () => {
        it("should match the snapshot for complex styled-component", async () => {
            const input = `
import styled from "styled-components";

const ComplexButton = styled.button\`
    color: \${props => props.color || "black"};
    background: \${props => props.primary ? "blue" : "gray"};
    border: none;
    padding: 10px 20px;
    border-radius: 5px;
    cursor: pointer;
    &:hover {
        background: \${props => props.primary ? "darkblue" : "darkgray"};
    }
    @media (max-width: 600px) {
        width: 100%;
        padding: 15px 25px;
    }
\`;
`;

            const result = await formatTemplateLiterals(input);
            expect(result).toMatchSnapshot();
        });
    });
});
