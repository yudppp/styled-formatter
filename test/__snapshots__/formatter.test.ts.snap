// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formatTemplateLiterals Snapshot tests should match the snapshot for complex styled-component 1`] = `
"
import styled from "styled-components";

const ComplexButton = styled.button\`
  color: \${props => props.color || "black"};
  background: \${props => props.primary ? "blue" : "gray"};
  border: none;
  padding: 10px 20px;
  border-radius: 5px;
  cursor: pointer;
  &:hover {
    background: \${props => props.primary ? "darkblue" : "darkgray"};
  }
  @media (max-width: 600px) {
    width: 100%;
    padding: 15px 25px;
  }
\`;
"
`;
